sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -91,14 +91,37 @@ static void *_event_loop(void *arg)
 
         if (res > 0) {
             switch (msg_rcvd.type) {
-                case GCOAP_MSG_TYPE_TIMEOUT:
-                    _expire_request((gcoap_request_memo_t *)msg_rcvd.content.ptr);
-                    break;
-                case GCOAP_MSG_TYPE_INTR:
-                    /* next _listen() timeout will account for open requests */
-                    break;
-                default:
-                    break;
+            case GCOAP_MSG_TYPE_TIMEOUT: {
+                gcoap_request_memo_t *memo = (gcoap_request_memo_t *)msg_rcvd.content.ptr;
+
+                /* no retries remaining */
+                if (memo->send_limit == GCOAP_SEND_LIMIT_NON
+                        || memo->send_limit == 0) {
+                    _expire_request(memo);
+                }
+                /* retries remaining; double timeout and resend */
+                else {
+                    unsigned i       = COAP_MAX_RETRANSMIT - memo->send_limit + 1;
+                    uint32_t timeout = ((uint32_t)COAP_ACK_TIMEOUT << i) * US_PER_SEC;
+                    timeout = random_uint32_range(timeout, timeout * COAP_RANDOM_FACTOR);
```

Comment:
mhm, I don't like that, bc `COAP_RANDOM_FACTOR` is `1.5` thereby introducing floating point maths here. We know that `timeout` is a multiple of `COAP_ACK_TIMEOUT` (which is two seconds). Hence, I would rather do something like 

```
#define COAP_TIMEOUT_VARIANCE    (500U * US_PER_MS)

uint32_t timeout = ((uint32_t)COAP_ACK_TIMEOUT << i) * US_PER_SEC;
timeout = random_uint32_range(timeout - COAP_TIMEOUT_VARIANCE, timeout + COAP_TIMEOUT_VARIANCE);
```